internal/storage/repository/repository.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/ajugalushkin/url-shortener-version2/internal/database"
+	"github.com/ajugalushkin/url-shortener-version2/internal/dto"
+	url_errors "github.com/ajugalushkin/url-shortener-version2/internal/errors"
+	"github.com/ajugalushkin/url-shortener-version2/internal/logger"
+	"github.com/jackc/pgerrcode"
+	"github.com/jackc/pgx/v5/pgconn"
+	_ "github.com/jackc/pgx/v5/stdlib"
+	"github.com/jmoiron/sqlx"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+func NewRepository(db *sqlx.DB) *Repo {
+	return &Repo{
+		db: db,
+	}
+}
+
+type Repo struct {
+	db *sqlx.DB
+}
+
+func (r *Repo) Put(ctx context.Context, shorteningInput dto.Shortening) (*dto.Shortening, error) {
+	var err error
+	err = database.WithTx(ctx, r.db, func(ctx context.Context, tx *sqlx.Tx) error {
+		sb := squirrel.StatementBuilder.
+			Insert("shorten_urls").
+			Columns("short_url", "correlation_id", "original_url").
+			PlaceholderFormat(squirrel.Dollar).
+			RunWith(r.db)
+
+		sb = sb.Values(
+			shorteningInput.ShortURL,
+			shorteningInput.CorrelationID,
+			shorteningInput.OriginalURL,
+		)
+
+		_, err = sb.ExecContext(ctx)
+		return err
+	})
+
+	if err != nil {
+		if pgErr, ok := errors.Unwrap(errors.Unwrap(err)).(*pgconn.PgError); ok && pgErr.Code == pgerrcode.UniqueViolation {
+			shortening, _ := r.GetByURL(ctx, shorteningInput.OriginalURL)
+			if shortening.OriginalURL != "" {
+				return shortening, &url_errors.DuplicateURLError{
+					Shortening: *shortening,
+					URLError:   fmt.Errorf("duplicate for %s", shortening.OriginalURL),
+				}
+			}
+		}
+		return nil, errors.Wrap(err, "repository.Put")
+	}
+	return &shorteningInput, nil
+}
+
+func (r *Repo) Get(ctx context.Context, shortURL string) (*dto.Shortening, error) {
+	var shorteningList []dto.Shortening
+
+	err := database.WithTx(ctx, r.db, func(ctx context.Context, tx *sqlx.Tx) error {
+		sb := squirrel.Select("short_url", "correlation_id", "original_url").
+			From("shorten_urls").
+			PlaceholderFormat(squirrel.Dollar).
+			Where(squirrel.Eq{"short_url": []string{shortURL}}).
+			RunWith(r.db)
+
+		query, args, err := sb.ToSql()
+		if err != nil {
+			return err
+		}
+
+		return r.db.SelectContext(ctx, &shorteningList, query, args...)
+	})
+	log := logger.LogFromContext(ctx)
+	if err != nil {
+		log.Info("repository.Get", zap.Error(err))
+		return nil, errors.Wrap(err, "repository.Get")
+	}
+
+	if len(shorteningList) == 0 {
+		log.Info("repository.Get", zap.Error(sql.ErrNoRows))
+		return nil, errors.Wrap(sql.ErrNoRows, "repository.Get")
+	}
+
+	shortening := shorteningList[0]
+
+	log.Info("repository.Get OK", zap.String("Original URL", shortening.OriginalURL))
+
+	return &shortening, nil
+}
+
+func (r *Repo) GetByURL(ctx context.Context, originURL string) (*dto.Shortening, error) {
+	var shorteningList []dto.Shortening
+
+	err := database.WithTx(ctx, r.db, func(ctx context.Context, tx *sqlx.Tx) error {
+		sb := squirrel.Select("short_url", "correlation_id", "original_url").
```

Comment:
CLARIFICATION: please build an execution cost plan of this query

Reply:
![image](https://github.com/ajugalushkin/url-shortener-version2/assets/50012581/177ece32-2991-43e4-8129-3570a8b78597)
If a specific URL is specified, I don't see a problem with the request